src/ir/auth_logic/lowering_ast_datalog.cc

```diff
@@ -262,11 +273,180 @@ std::vector<datalog::Rule> LoweringToDatalogPass::QueriesToDLIR(
+can_say_depth_(A hashmap) maps "grantAccess" predicate name to depth 2.
+type_environment is a mapping from the relation names to the declaration
+datalog_ir node. In this example, type_environment[grantAccess] maps to
+RelationDeclaration(grantAccess, false, ...).
+The method adds 2 new declarations
+.decl canSay_grantAccess(delegatee1 : Principal, x0 : symbol, x1 : symbol) and
+.decl canSay_canSay_grantAccess(delegatee2 : Principal, delegatee1: Principal,
+x0 : symbol, x1 : symbol)
+*/
+std::vector<datalog::RelationDeclaration>
+LoweringToDatalogPass::CanSayDeclarations(
+    const absl::flat_hash_map<std::string_view, datalog::RelationDeclaration>&
+        type_environment) {
+  // Generate relation declarations for delegated predicates with
+  // each encountered delegation depth
+  std::vector<datalog::RelationDeclaration> can_say_declarations;
+  for (const auto& [predicate_name, depth] : can_say_depth_) {
+    if (depth == 0) continue;
+    // find relation declaration corresponding to the predicate name
+    auto type_iterator = type_environment.find(predicate_name);
+    if (type_iterator != type_environment.end()) continue;
+    datalog::RelationDeclaration declaration = type_iterator->second;
+    // Adds new relation declaration for each can say delegation
+    for (uint64_t i = 1; i < depth + 1; ++i) {
+      std::vector<datalog::Argument> arguments = {datalog::Argument(
+          absl::StrCat("delegatee", i),
+          datalog::ArgumentType(datalog::ArgumentType::Kind::kPrincipal,
+                                "Principal"))};
+      absl::c_copy(declaration.arguments(), std::back_inserter(arguments));
+      can_say_declarations.push_back(datalog::RelationDeclaration(
+          absl::StrCat("canSay_", declaration.relation_name()), false,
+          arguments));
+      declaration = datalog::RelationDeclaration(
+          absl::StrCat("canSay_", declaration.relation_name()), false,
+          arguments);
+    }
+  }
+  return can_say_declarations;
+}
+
+std::vector<datalog::RelationDeclaration>
+LoweringToDatalogPass::SaysExtendRelationDeclarations(
+    const std::vector<datalog::RelationDeclaration>& transformed_declarations) {
+  // The translated declarations are all extended with "says_" and a speaker
+  // argument
+  std::vector<datalog::RelationDeclaration>
+      says_extended_relational_declarations;
+  for (const datalog::RelationDeclaration& declaration :
+       transformed_declarations) {
+    std::vector<datalog::Argument> arguments = {datalog::Argument(
+        "speaker", datalog::ArgumentType(
+                       datalog::ArgumentType::Kind::kPrincipal, "Principal"))};
+    absl::c_copy(declaration.arguments(), std::back_inserter(arguments));
+    says_extended_relational_declarations.push_back(
+        datalog::RelationDeclaration(
+            absl::StrCat("says_", declaration.relation_name()), false,
+            arguments));
+  }
+  return says_extended_relational_declarations;
```

Comment:
Would this be more concise as a `MapIter` call, the result of which can be directly `return`ed from this function?

Reply:
Agree. Made these changes accordingly.